Src/IronPython/Runtime/WeakRef.cs

```diff
@@ -91,67 +94,133 @@ struct CallbackInfo {
         }
 
         public void ChainCallback(object callback, object weakRef) {
-            _callbacks.Add(new CallbackInfo(callback, weakRef));
+            _lock.EnterWriteLock();
```

Comment:
Can we not lock on the _callbacks directly?

Reply:
locking on callbacks would be slightly less efficient - the read write lock allows simultaneous read access but synchronized write access whereas a simple lock will synchronize on all access...